src/main/kotlin/org/rust/lang/core/RustPsiPattern.kt

```diff
@@ -0,0 +1,35 @@
+import com.intellij.patterns.PatternCondition
+import com.intellij.psi.PsiComment
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiErrorElement
+import com.intellij.psi.PsiWhiteSpace
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.util.ProcessingContext
+import org.rust.lang.core.psi.RustTokenElementTypes
+
+/**
+ * Rust PSI tree patterns.
+ */
+class RustPsiPattern {
```

Comment:
Great, I didn't notice that. Though I think it would be better to move those patterns into a more general class than `AttributeCompletionProvider` if we want to reuse them. What do you think about `RustPsiPattern`? Would it be a suitable place for them?

I don't yet know how to properly unit test them in isolation, but I'll figure something out.